src/EFCore.PG/Storage/Internal/Mapping/NpgsqlHstoreTypeMapping.cs

```diff
@@ -55,28 +41,18 @@ protected override string GenerateNonNullSqlLiteral(object value)
             sb.Append('\'');
             return sb.ToString();
         }
-
-        class HstoreComparer : ValueComparer<Dictionary<string, string>>
+        protected static bool Compare(IReadOnlyDictionary<string, string> a, IReadOnlyDictionary<string, string> b)
         {
-            public HstoreComparer() : base(
-                (a, b) => Compare(a,b),
-                o => o.GetHashCode(),
-                o => o == null ? null : new Dictionary<string, string>(o))
-            {}
-
-            static bool Compare(Dictionary<string, string> a, Dictionary<string, string> b)
-            {
-                if (a == null)
-                    return b == null;
-                if (b == null)
+            if (a == null)
```

Comment:
I don't think this same logic applies for readonly/immutable: the way the code currently, runs, we needless compare all keys and values even if there was no change.

For readonly/immutable, we should just do a reference comparison and that's it. Technically, if the reference equality fails, we could still do a value comparison in case the dictionary was modified and then modified back - but that would penalize the 99% case where a different reference means different contents, for no real value (i.e. save a very rare database update that could have been prevented).